src/io_callbacks.zig

```diff
@@ -0,0 +1,902 @@
+                },
+                .close => |op| {
+                    linux.io_uring_prep_close(sqe, op.fd);
+                },
+                .connect => |*op| {
+                    linux.io_uring_prep_connect(sqe, op.socket, &op.address, op.address_size);
+                },
+                .fsync => |op| {
+                    linux.io_uring_prep_fsync(sqe, op.fd, op.flags);
+                },
+                .openat => |op| {
+                    linux.io_uring_prep_openat(sqe, op.fd, op.path, op.flags, op.mode);
+                },
+                .read => |op| {
+                    linux.io_uring_prep_read(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+                .recv => |op| {
+                    linux.io_uring_prep_recv(sqe, op.socket, op.buffer, op.flags);
+                },
+                .send => |op| {
+                    linux.io_uring_prep_send(sqe, op.socket, op.buffer, op.flags);
+                },
+                .write => |op| {
+                    linux.io_uring_prep_write(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+            }
+            sqe.user_data = @ptrToInt(completion);
+        }
+
+        fn complete(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
+                        os.EINTR => {
+                            completion.io.enqueue(completion);
+                            return;
+                        },
+                        os.EAGAIN => error.WouldBlock,
+                        os.EBADF => error.FileDescriptorInvalid,
+                        os.ECONNABORTED => error.ConnectionAborted,
+                        os.EFAULT => unreachable,
+                        os.EINVAL => error.SocketNotListening,
+                        os.EMFILE => error.ProcessFdQuotaExceeded,
+                        os.ENFILE => error.SystemFdQuotaExceeded,
+                        os.ENOBUFS => error.SystemResources,
+                        os.ENOMEM => error.SystemResources,
+                        os.ENOTSOCK => error.FileDescriptorNotASocket,
+                        os.EOPNOTSUPP => error.OperationNotSupported,
+                        os.EPERM => error.PermissionDenied,
+                        os.EPROTO => error.ProtocolFailure,
+                        else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                    } else @intCast(os.socket_t, completion.result);
+                    completion.callback(completion.context, completion, &result);
+                },
+                .close => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
+                        os.EINTR => {}, // A success, see https://github.com/ziglang/zig/issues/2425.
```

Comment:
Should we do `0` here instead of the empty block to be more explicit? I think `{}` will implicitly be `0` (I think that's how Zig does `void` or `u0`?) but perhaps we should go directly to `0`?

Reply:
`{}` (an empty block) has type `void` in zig, which is the type we want on a success for result which is a `CloseError!void` error union. Using `0` here wouldn't compile.

Reply:
That's perfect then. I don't know where I got the notion that `close()` should have a result of 0 instead of void!